Add OpenProject::Notifications as publish/subscribe mechanism

mfrister opened the pull request:
https://openproject.org/work_packages/5311

This allows code to register for certain events, we might later use
this later e.g. for sending mails on specific events like WorkPackage
updates.

Changelog:

```
* `#5311` Encapsulate ActiveSupport::Notifications
```

mfrister committed:
Add OpenProject::Notifications as publish/subscribe mechanism

This allows code to register for certain events, we might later use
this later e.g. for sending mails on specific events like WorkPackage
updates.

tessi commented on lib/open_project/notifications.rb:
```diff
@@ -0,0 +1,57 @@
+      ActiveSupport::Notifications.instrument(name, payload)
+    end
+
+    # Unsubscribe from a subscription as returned from .subscribe
```
This is hard to do, if the return value of `subscribe` should not be used. Maybe this should not belongt to the public API of the module.

mfrister replied:
With "opaque" I mean "you should not look into it or call anything on it, but only pass it to unsubscribe", not "don't use it", so it should only be used as an identification for the subscription.

Should I clarify that in the comment?
We need unsubscribe at least for tests, but we could remove it from public API. But then, I don't like changing the API only for making tests work.

tessi replied:
Tests can always use object.send(:unsubscribe) to call private methods. On the other hand we probably want to offer unsubscription at runtime. So it might be OK.

OK, ignore my last comment and youre code is good :)

tessi replied:
I'd say we'll talk on Monday about this - I need some :beers: now :)

mfrister replied:
Removed unsubscribe as it's not strictly necessary for tests. We modify global state in the spec, but we can prefix the notification names with the test name and I think having a more restricted interface is more valuable than properly cleaning up in a test.

Another argument for not providing unsubscribe is its [poor performance with our current ActiveSupport implementation](api.rubyonrails.org/v3.2.17/classes/ActiveSupport/Notifications.html):

> WARNING: The instrumentation framework is designed for long-running subscribers, use this feature sparingly because it wipes some internal caches and that has a negative impact on performance.

mfrister commented:
Also removed `name` and `id` as parameters from the registered block.

Removed name as we don't support registration to regexps, so the name is known at registration time. We don't have use for the `id` right now and we can later add it again as part of the payload hash.

tessi commented:
:thumbsup:
